btree/src/flatfile.rs

```diff
@@ -34,71 +41,103 @@ impl Appender {
+        let file = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(&filename)?;
 
-        let ahandle = {
-            let mut ahandle = fs::OpenOptions::new().append(true).open(&filename)?;
-            ahandle.seek(SeekFrom::End(0))?;
-            ahandle
-        };
-
-        let mut buf = [0u8; MAGIC_SIZE];
-        rhandle.read_exact(&mut buf)?;
-        rhandle.seek(SeekFrom::Start(DATA_START))?;
+        let storage = MmapStorage::new(file)?;
+        let next_pos = storage.len();
 
-        if buf != MAGIC {
-            return Err(Error::new(ErrorKind::Other, "magic mismatch"));
+        unsafe {
+            if storage.get(0, MAGIC_SIZE as u64) != MAGIC {
+                return Err(Error::new(ErrorKind::Other, "magic mismatch"));
+            }
         }
 
-        Ok(Self { rhandle, ahandle })
+        Ok(Self {
+            storage: RwLock::new(storage),
+            next_pos: AtomicU64::new(next_pos),
+        })
     }
 
     /// Check if this appender can still be appended to.
     pub fn can_append(&mut self) -> Result<bool, io::Error> {
-        let pos = self.ahandle.seek(SeekFrom::Current(0))?;
+        let pos = self.next_pos.load(Ordering::SeqCst);
         Ok(pos <= MAX_POS_OFFSET)
     }
 
     /// Append a blob of data and return the file offset
     ///
     /// Can only append data of MAX_BLOB_SIZE
-    pub fn append(&mut self, buf: &[u8]) -> Result<Pos, io::Error> {
+    pub fn append(&self, buf: &[u8]) -> Result<Pos, io::Error> {
         if buf.len() > MAX_BLOB_SIZE {
             return Err(Error::new(ErrorKind::Other, "blob size too big"));
         }
-        // if (buf.len() & 0b11) != 0 {
-        //     return Err(Error::new(
-        //         ErrorKind::Other,
-        //         "blob size is not a multiple of 4",
-        //     ));
-        // }
-        let pos = self.ahandle.seek(SeekFrom::Current(0))?;
-        if pos > MAX_POS_OFFSET {
+        if (buf.len() & 0b11) != 0 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                "blob size is not a multiple of 4",
+            ));
+        }
+
+        let next_pos = self.next_pos.load(Ordering::Acquire);
+        let mut storage = self.storage.upgradable_read();
+
+        if next_pos > MAX_POS_OFFSET {
             return Err(Error::new(ErrorKind::Other, "offset position too big"));
         }
+
         let blen = buf.len() as u32;
         let szbuf = blen.to_le_bytes();
-        self.ahandle.write_all(&szbuf[..])?;
-        self.ahandle.write_all(buf)?;
+
+        let region_len = szbuf.len() as u64 + buf.len() as u64;
+
+        let mmaped_region = unsafe {
+            match storage.get_mut(next_pos, region_len) {
+                Ok(slice) => slice,
+                Err(including) => {
+                    {
+                        let mut new_guard = RwLockUpgradableReadGuard::upgrade(storage);
+                        new_guard.extend(including)?;
```

Comment:
I've been thinking that may be it's just easier to just map a huge area from the start (basically enough to cover `MAX_POS_OFFSET`) and never resize. Was that the idea behind that constraint @vincenthz ? Although the only real advantage is that we could return pointers to the mapped are without the need to invalidate pointers/wait for readers to end before extending.

Reply:
you can only do that if you prereserve your file on disk, and it's probably unreasonable to pre-reserve a huge chunk (a terabyte), but it's not unreasonable to work on extend of around 128mb ... i.e. you can guarantee to mmap in a stable way 128mb at a time.

Reply:
isn't 128mb too conservative? right now it's growing exponentially, because I'm using powers of two, but I thought about setting a threshold and expand linearly from there. What I'm mostly wondering though, there is actually no way of working directly on the mmaped area, because each time you expand you invalidate pointers, so cloning is the only option, right?

Reply:
my idea of working on this is a pagetable of X size. the pagetable contains where a range of X size is mapped by holding its mmaped pointer (or null if not currently mapped).

Each time the file is grown, the file is extended by X through an `fallocate` or `ftruncate` and the pagetable is extended by 1.

Reply:
Oh, glad I asked then. So your idea was not to create a big contiguous mmap covering the whole file, but have something more granular and have something like a translation unit? I guess then you can map pages on demand, the only problem is deciding when to unmap them (if we actually do).

I'm not sure why didn't I consider doing something like that, although it is a bit more complex than what's done now.

As an aside note, that would probably make deleting things more easier.